crates/lib/src/qpu/experimental/random.rs

```diff
@@ -0,0 +1,272 @@
+/// convertible to `f64`.
+#[derive(Debug, Clone, Copy)]
+pub struct PrngSeedValue {
+    as_u64: u64,
+    pub(super) as_f64: f64,
+}
+
+impl PrngSeedValue {
+    /// Attempt to create a new instance of `PrngSeedValue` from a `u64`.
+    /// The value must be in the range `[1, MAX_SEQUENCER_VALUE]` and
+    /// losslessly convertible to `f64`.
+    pub fn try_new(value: u64) -> Result<Self> {
+        if !(1..=MAX_SEQUENCER_VALUE).contains(&value) {
+            return Err(Error::InvalidSeed(value));
+        }
+        if let Some(f64_value) = value.to_f64() {
+            Ok(Self {
+                as_u64: value,
+                as_f64: f64_value,
+            })
+        } else {
+            Err(Error::InvalidSeed(value))
+        }
+    }
+}
+
+fn lfsr_next(seed: u64, taps: &[u32]) -> u64 {
+    let feedback_value = taps.iter().fold(0, |acc, tap| {
+        let base = 2u64.pow(*tap - 1);
+        let bit = u64::from((seed & base) != 0);
+        acc.bitxor(bit)
+    });
+    (seed << 1) & MAX_SEQUENCER_VALUE | feedback_value
+}
+
+/// This represents the LFSR currently implemented on Rigetti control systems. Specifically,
+/// it implements a 48-bit LFSR with taps at indices 48, 47, 21, and 20.
+#[must_use]
+pub fn lfsr_v1_next(seed: PrngSeedValue) -> u64 {
+    lfsr_next(seed.as_u64, &[48, 47, 21, 20])
+}
+
+fn generate_lfsr_v1_sequence(seed: u64, start_index: u32, series_length: u32) -> Vec<u64> {
+    let mut lfsr = seed & MAX_SEQUENCER_VALUE;
+
+    let range = start_index..(start_index + series_length);
+    let mut collection = vec![];
+    for i in 0..(start_index + series_length) {
+        lfsr = lfsr_next(lfsr, &[48, 47, 21, 20]);
+        if range.contains(&i) {
+            collection.push(lfsr);
+        }
+    }
+    collection
+}
+
+fn prng_value_to_sub_region_index(value: u64, sub_region_count: u8) -> u8 {
+    ((value & MAX_UNSIGNED_MULTIPLIER) % u64::from(sub_region_count))
+        .to_u8()
+        .expect("modulo u8 should always produce a valid value")
+}
+
+/// Given a seed, start index, series length, and sub-region count, this function
+/// will generate the sequence of pseudo-randomly chosen indices Rigetti control
+/// systems.
+///
+/// For instance, if the following Quil program is run for 100 shots:
+///
+/// ```quil
+/// # presumed sub-region size is 3.
+/// DECLARE destination REAL[6] # prng invocations per shot = (6 / sub_region_size)  = 2
+/// DELCARE source REAL[12]     # implicit sub-region count = (12 / sub_region_size) = 4
+/// DECLARE seed INTEGER[1]
+/// DECLARE ro BIT[1]
+///
+/// DELAY 0 1e-6
+///
+// PRAGMA EXTERN choose_random_real_sub_regions "(destination : mut REAL[], source : REAL[], sub_region_size : INTEGER, seed : mut INTEGER)"
+// CALL choose_random_real_sub_regions destination source 3 seed
+/// ```
+///
+/// with a seed of 639523, you could backout the randomly chosen sub-regions with the following:
+///
+/// ```rust
+/// let seed = 639523;
```

Comment:
Add a `use` statement for `choose_random_real_sub_region_indices`. If this code should not fail while running, it'll be a successful doctest. Otherwise, replace `rust` with `no_run` so it'll do compilation without running

Reply:
Is it `use super:: choose_random_real_sub_region_indices`?

Reply:
It would be `qcs` instead of `crate`, but yes. Basically, write the import like it's being used in a separate crate.